uni/lib/view/restaurant/widgets/restaurant_page_card.dart

```diff
@@ -43,13 +47,50 @@ class CardFavoriteButton extends StatelessWidget {
       builder: (context, restaurantProvider) {
         final isFavorite =
             restaurantProvider.favoriteRestaurants.contains(restaurant.name);
+        final favoriteCardTypes =
+            Provider.of<HomePageProvider>(context).favoriteCards;
         return IconButton(
           icon: isFavorite ? Icon(MdiIcons.heart) : Icon(MdiIcons.heartOutline),
-          onPressed: () => restaurantProvider.toggleFavoriteRestaurant(
-            restaurant.name,
-          ),
+          onPressed: () => {
+            restaurantProvider.toggleFavoriteRestaurant(
+              restaurant.name,
+            ),
+            if (!isFavorite &&
+                !favoriteCardTypes.contains(FavoriteWidgetType.restaurant) &&
+                restaurantProvider.favoriteRestaurants.length < 2)
```

Comment:
Shouldn't we check if the card is actually present instead? So `if !isFavorite && !cardIsPinned`